PWGHF: implement event selection in track-index skim creator

fgrosa opened the pull request:
Add workflow that tags events and store information in a table with a single bitmap used to filter the collisions in the workflow for building the candidates

fgrosa committed:
Add workflow for event tagging

fgrosa committed:
Remove placeholder for event selection

fgrosa committed:
Add missing header

fgrosa committed:
Add histo to monitor selected/rejected events

fgrosa committed:
Fix format

aalkin commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,6 +33,47 @@ using namespace o2::aod;
+    //TODO: add more event selection criteria
+
+    // selected events
+    if (b_dovalplots) {
```
This is called for all events, not only selected, is it not?

fgrosa replied:
Right, thanks for spotting this! I also added a bin to monitor the number of processed events

aalkin commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -910,6 +943,7 @@ struct HFTrackIndexSkimsCreator {
 WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
 {
   return WorkflowSpec{
+    adaptAnalysisTask<SelectCollisions>(cfgc, TaskName{"hf-produce-sel-collision"}),
```
It would be better to just call the class `HfProduceSelCollisions` and avoid `TaskName` argument.

fgrosa replied:
Ok! Should we do it also for the other two classes?

aalkin replied:
Yes, please!

fgrosa replied:
Done! 👍

fgrosa committed:
Fix fill of event histo and change struct name

fgrosa committed:
Add missing if

fgrosa committed:
Change struct names to avoit TaskName argument

vkucera commented on Analysis/DataModel/include/AnalysisDataModel/HFSecondaryVertex.h:
```diff
@@ -26,6 +26,14 @@ using namespace o2::analysis;
 {
+namespace hf_selcollision
+{
+DECLARE_SOA_COLUMN(WhyRejectColl, whyRejectColl, int); //!
```
How many bits are needed here?

fgrosa replied:
I am not yet sure about that, in AliPhysics we had 16 bits (https://github.com/alisw/AliPhysics/blob/70706363d8eb5e911166640a9e83655ba7cf2d61/PWGHF/vertexingHF/AliRDHFCuts.h#L39) but I don't expect to have them all here (maybe others will be instead needed). My understanding however is that the event selection task does not have much information avaliable currently, so I only implemented the trigger class for the time being

ginnocen replied:
@fgrosa I would keep it as it is for the moment. It is difficult to foresee how much info we might want to include here.

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,8 +33,57 @@ using namespace o2::aod;
+  // event selection
+  void process(soa::Join<aod::Collisions, aod::EvSels>::iterator const& collision)
+  {
+    int status_collision = 0;
```
Is `status_collision` supposed to store other info than rejection reasons?

fgrosa replied:
No, only rejection reasons I would say

vkucera commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,8 +33,57 @@ using namespace o2::aod;
+
+  Produces<aod::HFSelCollision> rowSelectedCollision;
+
+  Configurable<bool> b_dovalplots{"b_dovalplots", true, "fill histograms"};
+  Configurable<std::string> s_trigger_class{"trigger_class", "kINT7", "trigger class"};
+  int trigger_class = std::distance(aliasLabels, std::find(aliasLabels, aliasLabels + kNaliases, s_trigger_class.value.data()));
+
+  HistogramRegistry registry{
+    "registry",
+    {{"h_events", "Events;;entries", {HistType::kTH1F, {{3, 0.5, 3.5}}}}}};
```
Fix the variable names. (camelCase, no underscores, no type prefixes, `h` in histogram names is fine)

fgrosa replied:
done in https://github.com/AliceO2Group/AliceO2/pull/6016/commits/6f14765ffdb8058b18276786f72da58a92d36d48. I used snake_case because it was used in the `HfProduceSelTrack` class, so maybe we could then modify this also there in a next iteration

fgrosa committed:
Use camelCase

fgrosa committed:
Use camelCase

ginnocen commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,8 +33,57 @@ using namespace o2::aod;
 using namespace o2::analysis::hf_cuts_single_track;
 
+/// Event selection
+struct HfProduceSelCollisions {
```
(Minor comment) I would call it HfTagSelCollisions, and similarly for the track select task. Just to avoid that one thinks that this creates track tables

fgrosa replied:
Changed

ginnocen commented on Analysis/Tasks/PWGHF/HFTrackIndexSkimsCreator.cxx:
```diff
@@ -32,8 +33,57 @@ using namespace o2::aod;
+    int statusCollision = 0;
+
+    if (doValPlots) {
+      registry.get<TH1>(HIST("hEvents"))->Fill(1);
```
I would move the histogram filling in a separate QAHFcollision utility, that can stay in this .cxx probably and be attached to the workflow. So that we keep this part of the code as clean as possibile

fgrosa replied:
Ok! For the moment I removed it, but I am working at the QA task in this other branch https://github.com/fgrosa/AliceO2/commit/454a6a57e8807fe0ae452bc90a56af4e53d4d046 (still work in progress)

ginnocen replied:
Then please leave it. We remove it in the next PR.

fgrosa replied:
Ok! I reverted it https://github.com/AliceO2Group/AliceO2/pull/6016/commits/65e28847d0adebc7e254ad1efd6e8093829b9620

ginnocen reviewed (commented):
@fgrosa all good on my side, apart a couple of minor comments. Please have a look. Thanks!

ginnocen commented:
hi @fgrosa please have a look, implement the comments if you think are useful and then flip it to ready for review. Thanks!

fgrosa committed:
Update name of  structs

fgrosa committed:
Remove output histogram

fgrosa committed:
Update workflows with new struct names

fgrosa committed:
Revert removal of histogram

ginnocen commented:
hi @iarsene @jgrosseo, this is good to go for me.

vkucera commented:
@ginnocen Merging this will most likely cause conflicts in Chiara's big PR: https://github.com/AliceO2Group/AliceO2/pull/5257
It might be better to wait and merge this after that.

ginnocen commented:
Ok @vkucera I will wait.

vkucera commented:
@fgrosa Please rebase and resolve conflicts.

fgrosa committed:
Fix conflicts

fgrosa committed:
Fix format

fgrosa commented:
> @fgrosa Please rebase and resolve conflicts.

Hi @vkucera I have fixed them, and I also added the event selection in the workflow for V0s added by @chiarazampolli (please have a look if it is fine with you)

ginnocen commented:
hi @fgrosa @vkucera are these errors connected to the PR? I would try to merge this PR asap and move forward with the hf track skim implementation. Cheers GM

fgrosa commented:
> hi @fgrosa @vkucera are these errors connected to the PR? I would try to merge this PR asap and move forward with the hf track skim implementation. Cheers GM

Hi @ginnocen indeed it looks to me that the errors are unrelated to the PR (they also change in time, sometimes is only build/O2/o2, sometimes both, sometimes none). Thanks!

ginnocen commented:
hi @iarsene @jgrosseo, could we merge this PR? the errors seem unrelated from what I understand. Thanks, GM

iarsene commented:
The fullCI and o2 fails are not related.
